Data/HashMap/Base.hs

```diff
@@ -443,17 +443,18 @@ instance (Hashable k) => H.Hashable1 (HashMap k) where
 #endif
 
 instance (Hashable k, Hashable v) => Hashable (HashMap k v) where
-    hashWithSalt salt hm = go salt (toList' hm [])
+    hashWithSalt salt hm = go salt hm
       where
-        go :: Int -> [HashMap k v] -> Int
-        go s [] = s
-        go s (Leaf _ l : tl)
-          = s `hashLeafWithSalt` l `go` tl
+        go :: Int -> HashMap k v -> Int
+        go s Empty = s
+        go s (BitmapIndexed _ a) = A.foldl' go s a
+        go s (Leaf h (L _ v))
+          = s `H.hashWithSalt` h `H.hashWithSalt` v
         -- For collisions we hashmix hash value
         -- and then array of values' hashes sorted
-        go s (Collision h a : tl)
-          = (s `H.hashWithSalt` h) `hashCollisionWithSalt` a `go` tl
-        go s (_ : tl) = s `go` tl
+        go s (Full a) = A.foldl' go s a
+        go s (Collision h a)
+          = (s `H.hashWithSalt` h) `hashCollisionWithSalt` a
```

Comment:
@treeowl This code makes sense to me, particular because we can be strict in the salt. Was there a reason we were lazily computing this before? If not, I'm 👍

Reply:
I don't think we were. Where do you see laziness?

Reply:
`toList'` seems to be implemented in terms of a `foldr`. It's probably a good exercise to go through and make sure it's not used in other cases where strict folds would be preferable.

Reply:
@emilypi there's no reason to prefer a strict fold for conversion to a list like that, and every reason not to.

Reply:
I think we're miscommunicating here. I'm not focused on list building - we removed that. I was wondering why it was being used where direct recursion with strict folds would have been better. It seems there are other places where that happens. I don't quite understand the introduction of `toList'` in many cases.